src/frontends/lean/decl_cmds.cpp

```diff
@@ -542,42 +542,45 @@ static void parse_equations_core(parser & p, buffer<expr> const & fns, buffer<ex
     while (true) {
         expr lhs;
         unsigned prev_num_undef_ids = p.get_num_undef_ids();
-        buffer<expr> locals;
-        {
-            parser::undef_id_to_local_scope scope2(p);
-            buffer<expr> lhs_args;
-            auto lhs_pos = p.pos();
-            if (p.curr_is_token(get_explicit_tk())) {
-                p.next();
-                name fn_name = p.check_decl_id_next("invalid recursive equation, identifier expected");
-                lhs_args.push_back(p.save_pos(mk_explicit(get_equation_fn(fns, fn_name, lhs_pos)), lhs_pos));
+        buffer<expr> lhs_args;
+
+        // check if lhs starts with an (explicit) equation function symbol (which is optional for fns.size() == 1)
+        optional<expr> fn;
+        auto lhs_pos = p.pos();
+        if (p.curr_is_token(get_explicit_tk())) {
+            p.next();
+            name fn_name = p.check_decl_id_next("invalid recursive equation, identifier expected");
+            lhs_args.push_back(p.save_pos(mk_explicit(get_equation_fn(fns, fn_name, lhs_pos)), lhs_pos));
+        } else if (p.curr_is_identifier() && (fn = is_equation_fn(fns, p.get_name_val()))) {
+            p.next();
+            lhs_args.push_back(p.save_pos(*fn, lhs_pos));
+        } else {
+            if (fns.size() == 1) {
+                lhs_args.push_back(p.save_pos(mk_explicit(fns[0]), lhs_pos));
             } else {
-                expr first = p.parse_expr(get_max_prec());
-                expr fn    = first;
-                if (is_explicit(fn))
-                    fn = get_explicit_arg(fn);
-                if (is_local(fn) && is_equation_fn(fns, local_pp_name(fn))) {
-                    lhs_args.push_back(first);
-                } else if (fns.size() == 1) {
-                    lhs_args.push_back(p.save_pos(mk_explicit(fns[0]), lhs_pos));
-                    lhs_args.push_back(first);
-                } else {
-                    throw parser_error("invalid recursive equation, head symbol in left-hand-side is not a constant",
-                                       lhs_pos);
-                }
+                throw parser_error("invalid recursive equation, head symbol in left-hand-side is not an equation function",
+                                   lhs_pos);
             }
+        }
```

Comment:
This refactoring turned out to be unnecessary after all, but I think it simplifies the function's logic